File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefProd = [0]*n
        suffProd = [0]*n
        prefProd[0] = 1
        suffProd[n-1] = 1

        for i in range(1,n):
            prefProd[i] = prefProd[i-1]*nums[i-1]
        
        for i in range(n-2,-1,-1):
            suffProd[i] = suffProd[i+1]*nums[i+1]
        
        ans = []
        for i in range(n):
            ans.append(prefProd[i]*suffProd[i])

        return ans

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.cpp
class Solution {
public:
    void helper(vector<int>& nums, int idx, vector<int>& ds, vector<vector<int>>& ans, int target){
        int n = nums.size();
        if(target == 0){
            ans.push_back(ds);
            return;
        }

        for(int i = idx; i < n; i++){
            if(i > idx && nums[i] == nums[i-1]) continue;
            if(nums[i] > target) break;
            ds.push_back(nums[i]);
            helper(nums,i+1,ds,ans,target-nums[i]);
            ds.pop_back();
        }

    }
    vector<vector<int>> combinationSum2(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        vector<int> ds;
        vector<vector<int>> ans;
        helper(nums,0,ds,ans,target);
        return ans;
    }
};
